### PR TITLE
[IP-280] Contract for IP claim sharing

### PR DESCRIPTION
**Pull request by vladichhh**

### CODE REVIEW

File: contracts/Share/IPClaimsShare.sol [hunk 1]
```diff
@@ -0,0 +1,74 @@
+pragma solidity ^0.4.23;
+
+import "../TestContracts/Token/ERC20.sol";
+
+
+contract IPClaimsShare {
```
Comment (thcrnk) [inline contracts/Share/IPClaimsShare.sol]: This does not extend ownableUpgradable. It won't work, try with unit tests

File: contracts/Share/IPClaimsShare.sol [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+import "../IPClaim/IPClaim.sol";
+
+
+contract IPClaimsShare is IIPClaimsShare, OwnableUpgradableImplementation {
+
+    using SafeMath for uint256;
+
+    uint256 public feeInUSD;
+    uint256 public feeInTokens;
+    address feeReceiver;
```
[the remark this answers is not in the document]
Comment (vladichhh) [reply contracts/Share/IPClaimsShare.sol]: fixed in last pull request

File: contracts/Share/IPClaimsShare.sol [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+import "../Oracle/ExchangeOracle.sol";
+import "../../node_modules/openzeppelin-solidity/contracts/math/SafeMath.sol";
+import "../IPClaim/IPClaim.sol";
+
+
+contract IPClaimsShare is IIPClaimsShare, OwnableUpgradableImplementation {
+
+    using SafeMath for uint256;
+
+    uint256 public feeInUSD;
```
[the remark this answers is not in the document]
Comment (vladichhh) [reply contracts/Share/IPClaimsShare.sol]: fixed in last pull request

File: contracts/Share/IPClaimsShare.sol [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+    address feeReceiver;
+    ERC20 public ipToken;
+    ExchangeOracle public exchangeOracle;
+
+    struct Recipient {
+        address addr;
+        bytes32 ndaHash;
+    }
+
+    mapping(bytes32 => bool) public sharedTo;
```
[the remark this answers is not in the document]
Comment (vladichhh) [reply contracts/Share/IPClaimsShare.sol]: fixed in last pull request

File: contracts/Share/IPClaimsShare.sol [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+
+    /**
+     * getters & setters
+     */
+
+    function getFeeInUSD() public view returns (uint256 _feeInUSD) {
+        return feeInUSD;
+    }
+
+    function setFeeInUSD(uint256 _feeInUSD) public onlyOwner onlyValidAmount(_feeInUSD) {
```
[the remark this answers is not in the document]
Comment (vladichhh) [reply contracts/Share/IPClaimsShare.sol]: removed modifier 'onlyValidAmount'
fixed in last pull request

File: contracts/Share/IPClaimsShare.sol [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+        // convert fee into wei
+        uint256 feeInWei = exchangeOracle.convertUSDToWei(feeInUSD);
+
+        // check the received ethers amount
+        require(msg.value >= feeInWei);
+
+        // transfer fee to the platform account
+        feeReceiver.transfer(feeInWei);
+
+        // refund if needed
```
[the remark this answers is not in the document]
Comment (vladichhh) [reply contracts/Share/IPClaimsShare.sol]: removed explicit refunding
fixed in last pull request

File: contracts/Share/IPClaimsShare.sol [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+
+        emit ExchangeOracleUpdated(_exchangeOracle);
+    }
+
+    /**
+     * external functions
+     */
+
+    function shareIPClaimWithEthers(address claim, address recipient, bytes32 ndaHash) external payable
+                onlyValidAddress(claim) onlyValidAddress(recipient) onlyValidHash(ndaHash)
```
[the remark this answers is not in the document]
Comment (vladichhh) [reply contracts/Share/IPClaimsShare.sol]: The difference without the 3 modifiers (checking for empty addresses) is ~ 1000 gas.

File: contracts/Share/IPClaimsShare.sol [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+                onlyClaimOwner(claim) isPrivate(claim) alreadyShared(claim, recipient) {
+
+        require(recipient != IPClaim(claim).owner(), "Can not share claim with its owner");
+
+        // convert fee into wei
+        uint256 feeInWei = exchangeOracle.convertUSDToWei(feeInUSD);
+
+        // check the received ethers amount
+        require(msg.value >= feeInWei);
```
[the remark this answers is not in the document]
Comment (vladichhh) [reply contracts/Share/IPClaimsShare.sol]: fixed in last pull request

**Pull request merged**